### DISCUSSION TITLE
API-Rules: Restrict the number of entries in a collection a user can create

### DISCUSSION THREAD

**Discussion by timstr97** [pocketbase/pocketbase#4442] [Q&A] [answered]
Hello,

i want to limit the entries a user can make in a collection based on the create API-Rule. When the user wants to create a new entry, the API-Rules should check if the user has surpassed the number of allowed entries.

I have the following data structure:

![grafik](https://github.com/pocketbase/pocketbase/assets/26687030/d57f9f52-250b-4209-882b-8ee6a67b77e4)

I would imagine a create rule at the vacancies-collection like this:

`@request.auth.business.allowedVacancies >= count(business.id)`

The `allowedVacancies` is a number saved for every business and describe the maximum vacancy-entries the business can make. The `count(business.id)` describes the number of vacancies of the business in the vacancies-collection. `employees` is a user-collection.

-------------------------------------------

I know that i could solve the problem if i change the data structure to this:

![grafik](https://github.com/pocketbase/pocketbase/assets/26687030/2dc5625a-bd89-4cde-872e-272673d02ec0)

Then it would work with the following update rule on the business-collection: 

`allowedVacancies >= @request.data.vacancies:length`

This second approach would lead to major refactorings. Is it possible to do it with the data structure of the first approach?

Thanks for the help

**Comment by ganigeorgiev** [comment]
_You can find similar discussions in https://github.com/pocketbase/pocketbase/discussions/3108, https://github.com/pocketbase/pocketbase/discussions/3513_.

In general if you want to use the data structure from the first approach you'll need to create a helper [View collection](https://pocketbase.io/docs/collections/#view-collection) with your specific count criteria that you'll then later join in the rules with `@collection.*` field.

For example (_I didn't test it_):

1. Create a View collection query with the aggregation for your specific use case (eg. `businessVacanciesCount`):
    ```sql
    SELECT
        v.business as id,
        v.business as business, -- add the column second time so that it can be used as relation field
        count("*") as totalVacancies
    FROM vacancies v
    GROUP BY v.business
    ```

2. Vacancies Create API rule:
   ```js
   @request.auth.id != "" &&
   // I don't fully understand your use case but maybe check that the user is creating a Vacancies record for their business 
   @request.data.business = @request.auth.business &&
   // join the view collection
   @collection.businessVacanciesCount.id ?= @request.auth.business &&
   // compare the counters
   // (we use <= because the Create rule execute AFTER a dry insert of the record)
   @collection.businessVacanciesCount.totalVacancies ?<= @request.auth.business.allowedVacancies
   ```
   _edit: updated the @collection.* matching to use `?=`_

**Reply by ganigeorgiev** [reply, marked as answer]
I forgot that with PocketBase v0.22.0 release an alternative to manually joining the view collection with the `@collection.*` field could be to use the new [back-relation](https://pocketbase.io/docs/working-with-relations/#back-relations) filter support:

```js
@request.auth.id != "" &&
@request.data.business = @request.auth.business &&
@request.auth.business.businessVacanciesCount_via_business.totalVacancies ?<= @request.auth.business.allowedVacancies
```

**Comment by timstr97** [comment]
Thanks a lot Gani, works like a charm! That was really helpful!
